### PR TITLE
Added resizing transformer and tutorial for use and problem it solves

### PR DESCRIPTION
**Pull request by DmitriyValetov**
Good day.

Added resizing transformer and tutorial that describes error that is caused by inequal sized time series + use of transformer.

#### PR checklist for new estimators 
- [x] I have added unit tests and made sure they pass locally. 
- [x] I have updated the existing example notebooks or provided a new one to showcase how my algorithm works.
- [ ] I have updated sktime's [estimator overview](https://github.com/alan-turing-institute/sktime/blob/master/ESTIMATOR_OVERVIEW.md) and I'm committed to maintain my contribution and provide bug fixes if necessary.

### CODE REVIEW

File: sktime/transformers/resizing.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            Series to user length via scipy interp1d between received points.
+    """
+
+    def __init__(self, length):
+        """
+        Parameters
+        ----------
+        length : integer, the length of time series to resize to.
+        """
+        assert(length>0)
```
Comment (mloning) [inline sktime/transformers/resizing.py:21]: Please replace assert statement with ValueError with some informative error message

File: sktime/transformers/resizing.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        self.length = length
+        super(TSResizeTransform).__init__()
+        
+    def __resizeCell(self, cell):
```
Comment (mloning) [inline sktime/transformers/resizing.py:25]: Avoid using [double underscores in Python](https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-a-single-and-a-double-underscore-before-an-object-name), private methods should have a single underscore. Please use [snake_case](https://en.wikipedia.org/wiki/Snake_case) for method names.

File: sktime/transformers/resizing.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+import pandas as pd
+from scipy import interpolate
+
+from sktime.transformers.base import BaseTransformer
+
+
+
+
```
Comment (mloning) [inline sktime/transformers/resizing.py:9]: remove empty lines, use some standard formatting tool

Comment (DmitriyValetov) [reply sktime/transformers/resizing.py]: done

File: sktime/transformers/resizing.py (hunk 2 above)
Comment (DmitriyValetov) [reply sktime/transformers/resizing.py]: done

File: sktime/transformers/resizing.py (hunk 1 above)
Comment (DmitriyValetov) [reply sktime/transformers/resizing.py]: done

**Comment by DmitriyValetov** [comment]
Test was added

File: sktime/tests/test_resizing.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import numpy as np
+import pandas as pd
+
+from sktime.datasets import load_basic_motions
+from sktime.transformers.resizing import TSResizeTransform
+
+# load data
+df, _ = load_basic_motions(split='TRAIN', return_X_y=True)
```
Comment (mloning) [inline sktime/tests/test_resizing.py:8]: please rename df to X as is standard for features

File: sktime/tests/test_resizing.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def cut_df_ts(df):
+    for row_i in range(df.shape[0]):
+        for dim_i in range(df.shape[1]):
+            ts = df.at[row_i, f'dim_{dim_i}']
+            df.at[row_i, f'dim_{dim_i}'] = pd.Series(ts.tolist()[:len(ts)-dim_i-1])
+
```
Comment (mloning) [inline sktime/tests/test_resizing.py:16]: please reformat files to delete unnecessary spaces

File: sktime/tests/test_resizing.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def test_resizing():
```
Comment (mloning) [inline sktime/tests/test_resizing.py:19]: this looks good, could you expand a little bit on the comments to explain how you're testing the transformer (just a few more words would already be quite helpful)? or making the code a bit more readable :-)

File: sktime/transformers/resizing.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        Parameters
+        ----------
+        X : nested pandas DataFrame of shape [n_samples, n_features]
+            Nested dataframe with time-series in cells.
+
+        Returns
+        -------
+        Xt : pandas DataFrame
+          Transformed pandas DataFrame with same number of rows and columns
+        """
```
Comment (mloning) [inline sktime/transformers/resizing.py:43]: one thing I forgot in my first review: could you use our [input validation function](https://github.com/alan-turing-institute/sktime/blob/f3c71977c113d986276fca2bdedf70faaa33f040/sktime/utils/validation/supervised.py#L9) for X

Comment (DmitriyValetov) [reply sktime/transformers/resizing.py:43]: done

File: sktime/tests/test_resizing.py (hunk 4 above)
Comment (DmitriyValetov) [reply sktime/tests/test_resizing.py:8]: done

File: sktime/tests/test_resizing.py (hunk 5 above)
Comment (DmitriyValetov) [reply sktime/tests/test_resizing.py:16]: Applyed autopep8 - is it sufficient?

File: sktime/tests/test_resizing.py (hunk 6 above)
Comment (DmitriyValetov) [reply sktime/tests/test_resizing.py:19]: done

**Comment by mloning** [comment]
I still have a few questions:

* What exactly happens if length is bigger than the length of the series in a cell? The series is stretched out to the required length?
* What exactly happens if length is smaller than the length of the series in a cell?
* If there is a time series (or pd.Series) in a cell, is the index also updated according to the interpolation? 
* Based on the answers to these questions, we should clarify the docstrings and perhaps rename it to make it clearer what it does

**Comment by DmitriyValetov** [comment]
1) If length is bigger than the length of the series in a cell - series is stretched.
2) If length is lower than the length of the series in a cell - series is sampled in points between the series points. If you have 100 point and you need to resize them to 50 points, you will receive 50 points with values: (s100[2*i]+s100[2*i+1])/2 for i in range(50).
3) I think, yes. Series is converted to list firstly and back to series. Don't we always have a series in a cell?

**Comment by mloning** [comment]
Thanks for the clarifications! 

* So the values of the series are interpolated as you describe, but when you convert it into a list you loose the index, so the index is not interpolated, is that right? 
* Why do we work with lists and not numpy arrays? 
* Please update the docstrings accordingly to describe what it does

**Comment by DmitriyValetov** [comment]
2) Use of list is my bad habit - corrected.
1) Yes there will be no index, type of cell value after transformation will be numpy. ".apply()" function can't return pandas.Series objects. But I can make additional cell-wise cycle for changing type to pandas.Series.
3) Docstrings refilled.

File: sktime/transformers/resizing.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import numpy as np
+import pandas as pd
+from scipy import interpolate
+
+from sktime.transformers.base import BaseTransformer
+from sktime.utils.validation.supervised import validate_X
+
+
+class TSResizeTransform(BaseTransformer):
+    """Transformer that rescales series for another number of points.
```
Comment (mloning) [inline sktime/transformers/resizing.py:10]: Please update doc string:
* ensure you follow numpy/scikit-learn formatting
* add descriptions what happens when series is longer/shorter than length from your comment on PR
* the transformer doesn't "rescale" the series, it rather interpolates/subsamples it, right?

**Comment by mloning** [comment]
Okay, thanks for the updates!
* I've left a minor comment on the doc string. 
* We're currently undergoing some refactoring and I want to merge those changes first because they add a number of additional unit tests. Once that's done, it may require you to make some additional minor changes. After that I'm happy to merge your PR. I'll keep you posted! We're aiming for a new release at the end of the month.

File: sktime/transformers/resizing.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import numpy as np
+import pandas as pd
+from scipy import interpolate
+
+from sktime.transformers.base import BaseTransformer
```
Comment (mloning) [inline sktime/transformers/resizing.py:5]: Please inherit from `BaseSeriesToSeriesTransformer`

File: sktime/transformers/resizing.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        Parameters
+        ----------
+        X : nested pandas DataFrame of shape [n_samples, n_features]
+            Nested dataframe with time-series in cells.
+
+        Returns
+        -------
+        pandas DataFrame : Transformed pandas DataFrame with same number of rows and columns
+        """
+        validate_X(X)
```
Comment (mloning) [inline sktime/transformers/resizing.py:71]:
```suggestion
        check_X(X)
```

Comment (mloning) [reply sktime/transformers/resizing.py:71]: Does your transformer work with multiple columns? If not, please set

```python
check_X(X, enforce_univariate=True)
```

File: sktime/transformers/resizing.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class TSResizeTransform(BaseTransformer):
+    """Transformer that rescales series for another number of points. 
+        For each cell in datadrame transformer fits scipy linear interp1d 
+        and samples user defined number of points. Points are generated 
+        by numpy.linspace. After transformation each cell will be a numpy.array
+        of defined size.
+    """
+
```
Comment (mloning) [inline sktime/transformers/resizing.py:16]: please add `_required_parameter = ["length"]` as a class attribute
```suggestion
         _required_parameter = ["length"]
```

File: sktime/transformers/resizing.py (hunk 9 above)
Comment (DmitriyValetov) [reply sktime/transformers/resizing.py:4]: May be, BaseSingleSeriesTransformer from sktime.transformers.single_series.base?
Cant' find BaseSeriesToSeriesTransformer.

Comment (mloning) [reply sktime/transformers/resizing.py]: ah sorry, it's called [`BaseSeriesAsFeaturesTransformer`](https://github.com/alan-turing-institute/sktime/blob/dev/sktime/transformers/series_as_features/base.py)

**Comment by mloning** [comment]
Please move your files into the `sktime/transformers/series_as_features/ ` directory

File: sktime/tests/test_interpolate.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    # 2) cutting each time series in each cell of X to make lengths different
+    cut_X_ts(X)  # operation is inplace
+    # get lengths to ensure that they are really different
+    ts_lens_after_cut = [len(X.iloc[0][i]) for i in range(len(X.iloc[0]))]
+    assert(not all([length == ts_lens_after_cut[0]
+                    for length in ts_lens_after_cut]))  # are different
+
+    # 3) make tranformer, set target length `target_len` and apply it
+    target_len = 50
+    Xt = TSInterpolator(target_len).transform(X)
```
Comment (mloning) [inline sktime/tests/test_interpolate.py:39]:
```suggestion
    Xt = TSInterpolator(target_len).fit_transform(X)
```

**Comment by mloning** [comment]
@DmitriyValetov I fixed two minor things, hope that's okay. All tests seem to pass now, everything good to go from your side?

**Comment by mloning** [comment]
Merged now, thanks @DmitriyValetov for the contribution! 🎉

**Comment by mloning** [comment]
@DmitriyValetov the example notebook that came with this PR is broken, could you open a new PR to fix it? I've removed it for now.

**Comment by DmitriyValetov** [comment]
> 
> 
> @DmitriyValetov the example notebook that came with this PR is broken, could you open a new PR to fix it? I've removed it for now.

Done

**Comment by mloning** [comment]
@DmitriyValetov I can't see the PR, where did you push the changes?

**Pull request merged**